libs/common/include/helpers/serializeContainers.h

```diff
@@ -0,0 +1,120 @@
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Return To The Roots. If not, see <http://www.gnu.org/licenses/>.
+
+#pragma once
+
+#include "serializeEnums.h"
+#include "s25util/Serializer.h"
+#include <type_traits>
+
+namespace helpers {
+
+namespace detail {
+    template<typename T, typename = void>
+    struct isResizableContainer : std::false_type
+    {};
+    template<typename T>
+    struct isResizableContainer<T, decltype(std::declval<T>().resize(size_t{}))> : std::true_type
+    {};
+
+    template<typename T>
+    std::enable_if_t<std::is_enum<T>::value, T> popEnumOrIntegral(Serializer& ser)
+    {
+        return popEnum<T>(ser);
+    }
+    template<typename T>
+    std::enable_if_t<!std::is_enum<T>::value, T> popEnumOrIntegral(Serializer& ser)
+    {
+        return ser.Pop<T>();
+    }
+
+    template<typename T>
+    void pushContainer(Serializer& ser, const T& container, long)
+    {
+        using Type = typename T::value_type;
+        using Integral = typename std::conditional_t<std::is_enum<Type>::value, std::underlying_type<Type>,
+                                                     std::common_type<Type>>::type;
+        for(const auto el : container)
+        {
+            // Cast also required for bool vector -.-
+            ser.Push(static_cast<Integral>(el));
+        }
+    }
+    template<typename T>
+    auto pushContainer(Serializer& ser, const T& container, int) -> std::enable_if_t<sizeof(*container.data()) == 1u>
+    {
+        ser.PushRawData(container.data(), container.size());
+    }
+
+    // If the container is resizable, pop the size and resize it
+    template<typename T>
+    std::enable_if_t<isResizableContainer<T>::value> maybePopSizeAndResize(Serializer& ser, T& container)
+    {
+        container.resize(ser.PopVarSize());
+    }
+    template<typename T>
+    std::enable_if_t<!isResizableContainer<T>::value> maybePopSizeAndResize(Serializer&, T&)
+    {}
+
+    template<typename T>
+    void popContainer(Serializer& ser, T&& container, long)
+    {
+        using Type = typename std::remove_reference_t<T>::value_type;
+        for(auto&& el : container) // auto&& required for vector<bool> proxy object
+            el = popEnumOrIntegral<Type>(ser);
+    }
+    template<typename T>
+    auto popContainer(Serializer& ser, T& container, int)
+      -> std::enable_if_t<sizeof(*container.data()) == 1u && !std::is_enum<typename T::value_type>::value>
+    {
+        ser.PopRawData(container.data(), container.size());
+    }
+
+} // namespace detail
+
+template<typename T>
+void pushContainer(Serializer& ser, const T& container, bool ignoreSize = false)
+{
+    using Type = typename T::value_type;
+    static_assert(std::is_integral<Type>::value || std::is_enum<Type>::value,
+                  "Only integral types and enums are possible");
+
+    if(detail::isResizableContainer<T>::value && !ignoreSize)
```

Comment:
probably we should still always transfer the sizes, would still make it less error prone

Reply:
There are multiple reasons not to. For fixed size containers (std::array) it is not required at all and would just make the messages/savegames larger. I introduced the VarSize format to save some size. And sometimes we really don't want to, e.g. for the compatibility code where size is serialized differently in older versions
As for error prone: We have multiple safeguards in place already: After every Event and GameObject there is a marker as a kind of "checksum". That was enough to find almost all issues I introduced when implementing this. The only issue that was slightly harder (but still found by that) was a confusion of push and pop (we should really have 2 classes... but: later)

Reply:
I think the size is not the problem here, the question is if its recoverable/reconstructable if we change the sizes of the arrays or between platforms.

Reply:
>  if we change the sizes of the arrays 

This is the same as changing the types of the data or adding/removing any other field: That makes the serialization fail (latest at the next checksum) and needs to be handled explicitely at the call site as is currently done for e.g. the different gameDataVersions

> between platforms.

There is no platform dependency here, so this isn't an issue :)

Reply:
ok as long we're not running into any deserialization problems (alignement) when you move your savegame to a different platform